### PR TITLE
Implement operators logical_and and logical_or

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24379 Implement operators logical_and and logical_or**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24379

Differential Revision: [D16830128](https://our.internmc.facebook.com/intern/diff/D16830128)

### CODE REVIEW

**Comment by xuhdev** [comment]
@bddppq Are you aware of a `nvfunctional` equivalent in ROCm?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: test/test_torch.py
Comment (xuhdev) [reply test/test_torch.py]: OK. I fixed other comments. I'll stack the fix of this issue in another PR to reduce the complexity of this PR.

Comment (xuhdev) [reply test/test_torch.py]: Or, if @VitalyFedyunin has a plan to sweep down all these kinds of tests, I'll leave this intact.

Comment (xuhdev) [reply test/test_torch.py]: Resolved -- turns out to be a quite simple change. And this is really good work, @VitalyFedyunin !

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
Comment (xuhdev) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: I now changed this to templating. According to my experiments, nvcc seems to be able to inline functions passed in as template parameters.

```cuda
__host__ __device__ float called1(float x) {
  return x + 1;
}

__host__ __device__ float called2(float x) {
  return x + 10;
}

template <typename Op>
__host__ __device__ float caller(float x, Op op) {
  return op(x);
}

__global__ void user(float x, float& y) {
  x = caller(x, called1);
  y = caller(x, called2);
}
```

After `nvcc -ptx -src-in-ptx -arch=sm_60 test.cu`:

```
//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-24817639
// Cuda compilation tools, release 10.0, V10.0.130
// Based on LLVM 3.4svn
//

.version 6.3
.target sm_60
.address_size 64

	// .globl	_Z4userfRf

.visible .entry _Z4userfRf(
	.param .f32 _Z4userfRf_param_0,
	.param .u64 _Z4userfRf_param_1
)
{
	.reg .f32 	%f<4>;
	.reg .b64 	%rd<3>;


	ld.param.f32 	%f1, [_Z4userfRf_param_0];
	ld.param.u64 	%rd1, [_Z4userfRf_param_1];
	cvta.to.global.u64 	%rd2, %rd1;
	add.f32 	%f2, %f1, 0f3F800000;
	add.f32 	%f3, %f2, 0f41200000;
	st.global.f32 	[%rd2], %f3;
	ret;
}
```

**Comment by xuhdev** [comment]
Added `logical_or` as well.

For some reason, after this addition, `BinaryOpsKernel.cu` gets really slow to build now. Not really sure how to deal with this without losing features.

**Comment by fmassa** [comment]
Should we delete the previous implementation (in TH) and redirect `&` and `|` operators to `logical_and` and `logical_or` now?

**Comment by VitalyFedyunin** [comment]
> Should we delete the previous implementation (in TH) and redirect `&` and `|` operators to `logical_and` and `logical_or` now?

Not sure we can do it, as our current `_th_or` is bit-wise when this implementation cast everything to boolean before applying `or`.

**Comment by gchanan** [comment]
bitwise is what's used in numpy for those operators.

**Comment by xuhdev** [comment]
I think `&` and `|` correspond to `bitwise_and` and `bitwise_not` (not implemented).

**Review by gchanan — CHANGES_REQUESTED** [review]

**Comment by xuhdev** [comment]
@ezyang Do you have any suggestions for failures like the following in `caffe2_py2_cuda9_1_cudnn7_ubuntu16_04_build`? My speculation is that it is caused by the long time to compile BinaryOpsKernel.cu.

```
Aug 19 20:02:45 error: failed to execute compile
Aug 19 20:02:45 caused by: error reading compile response from server
Aug 19 20:02:45 caused by: Failed to read response header
Aug 19 20:02:45 caused by: failed to fill whole buffer
Aug 19 20:02:45 CMake Error at torch_generated_BinaryOpsKernel.cu.o.Release.cmake:280 (message):
Aug 19 20:02:45   Error generating file
Aug 19 20:02:45   /var/lib/jenkins/workspace/build/caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/./torch_generated_BinaryOpsKernel.cu.o
Aug 19 20:02:45 
Aug 19 20:02:45 
Aug 19 20:02:45 make[2]: *** [caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/torch_generated_BinaryOpsKernel.cu.o] Error 1
Aug 19 20:02:45 caffe2/CMakeFiles/torch.dir/build.make:1232: recipe for target 'caffe2/CMakeFiles/torch.dir/__/aten/src/ATen/native/cuda/torch_generated_BinaryOpsKernel.cu.o' failed
```

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/quantized/README.md
Comment (xuhdev) [reply aten/src/ATen/native/quantized/README.md]: I changed to `XAND`. Hopefully I've plucked all ANDs from this file now.

File: aten/src/ATen/native/BinaryOps.cpp
Comment (xuhdev) [inline aten/src/ATen/native/BinaryOps.cpp]: In my latest change, I changed `Stub` to `Stub&`, as I've noticed that a `Stub` object caches some results in its `()` operator and I think we should use an lvalue reference here, which shouldn't hurt even if it doesn't help. All other changes, other than those that addressed other issues, are merely rebasing.

Comment (gchanan) [reply aten/src/ATen/native/BinaryOps.cpp]: ah, ya, I was going to point that out -- I wonder if we should disable the copy constructor in the dispatch stubs to make it clear doing this is wrong.

Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: If it is clearly wrong, I think we should (in a separate PR)

Comment (gchanan) [reply aten/src/ATen/native/BinaryOps.cpp]: sounds good.

**Comment by gchanan** [comment]
hmm is it worth splitting up BinaryOpsKernel.cu?  I'd agree with you that the likely culprit is that it's taking too long to compile.

**Comment by xuhdev** [comment]
> hmm is it worth splitting up BinaryOpsKernel.cu? I'd agree with you that the likely culprit is that it's taking too long to compile.

Sure thing. Does it make sense to you if I split all three functions to three different files? While semantically weird, this seems to me the most reasonable solution as it looks like the compilation of these logical ops are dominant.

**Comment by gchanan** [comment]
good question; the first thing I'd try is just to break out all the logical ops into e.g. BinaryLogicalOps and see if that works (with a comment that you had to break it up for compilation speed reasons).  But I think this is mainly being driven by build/test constraints, not code hygiene, so as long as you find something that works and it's commented why you did it I think it's fine.

**Comment by xuhdev** [comment]
@gchanan I broke down the CUDA implementation now

**Comment by ezyang** [comment]
> Do you have any suggestions for failures like the following in caffe2_py2_cuda9_1_cudnn7_ubuntu16_04_build? My speculation is that it is caused by the long time to compile BinaryOpsKernel.cu.

In the past, we've found that reducing parallelism helps avoid this sccache race. https://github.com/mozilla/sccache/issues/248

Apparently upstream has fixed this? I wonder if we are on a recent enough version.

**Comment by xuhdev** [comment]
@ezyang At least the failed one still have an old version installed, dated back to mid 2018.

**Comment by gchanan** [comment]
until we upgrade, it looks like there are failures on the .cpp files and not the .cu files -- so perhaps if you do the same thing for the .cpp files things will work.

Also, could you rebase?

**Comment by xuhdev** [comment]
@gchanan Done!

**Comment by gchanan** [comment]
also I pinged @yf225 about upgrading sccache and he says it's blocked on upstreaming some nvcc support.

**Comment by yf225** [comment]
sccache upgrade work is tracked at https://github.com/pytorch/pytorch/issues/24982.

**Review by gchanan — APPROVED** [review]
This lgtm, thanks and sorry about all the CI issues...

If you are so inclined, two things that would be nice to do in a follow-on PR:
1) See if we can combine AND/OR/etc into one kernel .cpp/.cu file.  I bet it will work, but the maximum split you are doing seems reasonable to unblock this for now.
2) Instead of repeating the "The content of ... message", write it once in a `// [NOTE: ]` and just reference the note when it's relevant.  See e.g. https://github.com/pytorch/pytorch/blob/e6b0ebdfd5c41aa6ff5a0430b45fc141be34af9b/aten/src/TH/THTensor.hpp#L50-L56 and https://github.com/pytorch/pytorch/blob/e6b0ebdfd5c41aa6ff5a0430b45fc141be34af9b/aten/src/THC/generic/THCTensor.h#L24 as an example.

But I'll commit this as is (assuming tests pass).

**Comment by xuhdev** [comment]
1. I actually tested locally (all binary logical operators in one single ".cu" file) and it turned out that it generates a `*.gpu` of the same size as before and it took equally long time, that's why I didn't try that on the CI. For the CPU version, they mirror the layout of the CUDA version to ease future maintenance, but I think this can be changed if we don't want the mirroring.

2. I'll do this in a follow up PR.

**Comment by ezyang** [comment]
> See if we can combine AND/OR/etc into one kernel .cpp/.cu file. I bet it will work, but the maximum split you are doing seems reasonable to unblock this for now.

I'm pretty sure nvcc doesn't scale linearly with file size. So it is better to keep kernels in as small files as possible. We saw this with "THCEvenMoreMath.cu".

**Review by gchanan — CHANGES_REQUESTED** [review]
@colesbury calculated that these triple dispatches result in 1400+ kernels being instantiated, which means incremental build times are ~15 minutes.  So we need to figure out a different way of doing this.

**Review by colesbury — CHANGES_REQUESTED** [review]

**Comment by cpuhrsch** [comment]
It looks like there are enough eyes on this PR so I'm marking it as triaged.

**Pull request closed without merging**
